cpp/src/lists/contains.cu

```diff
@@ -33,308 +31,251 @@
+ */
+auto constexpr __device__ NULL_SENTINEL = std::numeric_limits<size_type>::min();
 
-auto get_search_keys_device_iterable_view(cudf::scalar const& search_key, rmm::cuda_stream_view)
+/**
+ * @brief Indicate the current supported types in `cudf::lists::contains`.
+ *
+ * TODO: Add supported nested types.
+ */
+template <typename Element>
+static auto constexpr is_supported_non_nested_type()
 {
-  return &search_key;
+  return cudf::is_fixed_width<Element>() || std::is_same_v<Element, cudf::string_view>;
 }
 
-template <typename ElementType, duplicate_find_option find_option>
-auto __device__ find_begin(list_device_view const& list)
-{
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return list.pair_rep_begin<ElementType>();
-  } else {
-    return thrust::make_reverse_iterator(list.pair_rep_end<ElementType>());
-  }
-}
+/**
+ * @brief Functor to perform searching for index of a key element in a given list.
+ */
+struct search_list_fn {
+  duplicate_find_option const find_option;
 
-template <typename ElementType, duplicate_find_option find_option>
-auto __device__ find_end(list_device_view const& list)
-{
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return list.pair_rep_end<ElementType>();
-  } else {
-    return thrust::make_reverse_iterator(list.pair_rep_begin<ElementType>());
-  }
-}
+  template <typename Element, CUDF_ENABLE_IF(is_supported_non_nested_type<Element>())>
+  __device__ size_type operator()(list_device_view list, thrust::optional<Element> key_opt) const
+  {
+    // A null list or null key will result in a null output row.
+    if (list.is_null() || !key_opt) { return NULL_SENTINEL; }
 
-template <duplicate_find_option find_option, typename Iterator>
-size_type __device__ distance([[maybe_unused]] Iterator begin, Iterator end, Iterator find_iter)
-{
-  if (find_iter == end) {
-    return absent_index;  // Not found.
+    return find_option == duplicate_find_option::FIND_FIRST
+             ? search_list<true, Element>(list, *key_opt)
+             : search_list<false, Element>(list, *key_opt);
   }
 
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return find_iter - begin;  // Distance of find_position from begin.
-  } else {
-    return end - find_iter - 1;  // Distance of find_position from end.
+  template <typename Element, CUDF_ENABLE_IF(cudf::is_nested<Element>())>
+  __device__ size_type operator()(list_device_view, thrust::optional<Element>) const
+  {
+    CUDF_UNREACHABLE("Nested types are not yet supported");
   }
-}
 
-/**
- * @brief __device__ functor to search for a key in a `list_device_view`.
- */
-template <duplicate_find_option find_option>
-struct finder {
-  template <typename ElementType>
-  __device__ size_type operator()(list_device_view const& list, ElementType const& search_key) const
+ private:
+  template <bool forward, typename Element, CUDF_ENABLE_IF(is_supported_non_nested_type<Element>())>
```

Comment:
SFINAE may not be required here since the calling code has it. But it is fine to have it too.